### DISCUSSION TITLE
All POST requests fail with An error occurred when parsing the HTTP request POST at &#39;/&#39;

### DISCUSSION THREAD

**Discussion by david-benes** [minio/minio#18546] [Q&A] [answered]
When attempting to perform presigned POST upload, all requests fail with HTTP 400 and the following response:

```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>BadRequest</Code><Message>An error occurred when parsing the HTTP request POST at &#39;/&#39;</Message><Resource>/</Resource><RequestId></RequestId><HostId></HostId></Error>
```

Since I'm new to `minio`, there is a high chance that I am doing something wrong. However, I am able to perform the upload to Digital Ocean Spaces with the same code.

These are parameters that I use to prepare form data:

```
{
    url: 'http://localhost:9000/',
    fields: {
      bucket: 'bucket-name',
      'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
      'X-Amz-Credential': 'minioAccessKey/20231127/local/s3/aws4_request',
      'X-Amz-Date': '20231127T195940Z',
      key: 'uploads/d5075acc-afb7-406b-88c4-9d22eb1d8d8c/71b497a2-a74e-4fb9-8d08-9e44c65251fd.docx',
      Policy:
        'eyJleHBpcmF0aW9uIjoiMjAyMy0xMS0yN1QyMDowOTo0MFoiLCJjb25kaXRpb25zIjpbeyJidWNrZXQiOiJoYXRjaGVkLWJ5LWFpLWxvY2FsIn0seyJYLUFtei1BbGdvcml0aG0iOiJBV1M0LUhNQUMtU0hBMjU2In0seyJYLUFtei1DcmVkZW50aWFsIjoibWluaW9BY2Nlc3NLZXkxLzIwMjMxMTI3L2xvY2FsL3MzL2F3czRfcmVxdWVzdCJ9LHsiWC1BbXotRGF0ZSI6IjIwMjMxMTI3VDE5NTk0MFoifSx7ImtleSI6InVwbG9hZHMvZDUwNzVhY2MtYWZiNy00MDZiLTg4YzQtOWQyMmViMWQ4ZDhjLzcxYjQ5N2EyLWE3NGUtNGZiOS04ZDA4LTllNDRjNjUyNTFmZC5kb2N4In1dfQ==',
      'X-Amz-Signature':
        '65c14c59e7a4a09b11a2fc4ee9dc16ded298291525acd887b64c34260790ec88',
    },
  }
```

However, regardless of what I include in the POST request, the error is always the same.
I simplified the reproduction to following command:

`curl -L localhost:9000 -X POST`

What could be important is that no log/error appears in `mc admin trace -v` for any POST request to minio.
Issuing `curl -L localhost:9000 -X GET` responds correctly with Access Denied error and adds relevant record to log.
It appears like if POST requests do not even reach minio.

When I issue `http://localhost:9000/bucket/test`, the error is:
```
<Message>An error occurred when parsing the HTTP request POST at &#39;/bucket/test&#39;</Message><Resource>/bucket/test</Resource>
```

## Expected Behavior
I would expect POST request to make a record in `mc admin trace -v`.
I would also expect the request to fail on Access Denied.
And I would expect that correcly prepared POST payload with presigned data would upload a file successfully. 

## Current Behavior
`An error occurred when parsing the HTTP request POST`

## Steps to Reproduce (for bugs)

1. Install minio in docker container
2. Run `curl -L localhost:9000 -X POST`

## Context
I'm trying to perform presigned POST upload. However, it seems that all POST requests fail.

## Regression
no

## Your Environment
* Version used: `RELEASE.2023-11-20T22-40-07Z (go1.21.4 linux/arm64)`
* Server setup and configuration: `localhost:9000, docker image`
* Operating System and version: `Darwin HOSTNAME.local 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:45 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6020 arm64`

**Comment by leoneltrich** [comment]
I am experiencing the same problem. Did you manage to find a solution?

**Reply by david-benes** [reply]
No really. I created a real bucket for local development.
However, I will reconsider this approach and try again with the solution proposed by @mitkins.

**Comment by ijoseph** [comment]
I'm able to get this `cURL` command to work:

```bash
curl -X POST http://127.0.0.1:9000/all \
 -F "x-amz-algorithm=AWS4-HMAC-SHA256" \
 -F "x-amz-credential=minioadmin/20240415/us-east-1/s3/aws4_request" \
 -F "x-amz-date=20240415T090019Z" \
 -F "policy=eyJleHBpcmF0aW9uIjogIjIwMjQtMDQtMTZUMDk6MDA6MTkuOTkzWiIsICJjb25kaXRpb25zIjogW1siZXEiLCAiJGJ1Y2tldCIsICJhbGwiXSwgWyJzdGFydHMtd2l0aCIsICIka2V5IiwgIk1hIl0sIFsiY29udGVudC1sZW5ndGgtcmFuZ2UiLCAxLCAxMDQ4NTc2MF0sIFsiZXEiLCAiJHgtYW16LWFsZ29yaXRobSIsICJBV1M0LUhNQUMtU0hBMjU2Il0sIFsiZXEiLCAiJHgtYW16LWNyZWRlbnRpYWwiLCAibWluaW9hZG1pbi8yMDI0MDQxNS91cy1lYXN0LTEvczMvYXdzNF9yZXF1ZXN0Il0sIFsiZXEiLCAiJHgtYW16LWRhdGUiLCAiMjAyNDA0MTVUMDkwMDE5WiJdXX0=" \
 -F "x-amz-signature=ad19470138791ff028fab00a88908ae2036cbb3fdcceafcab37e9a9fb9ccdc9d" \
 -F "Key=Makefile" \
 -F file=@Makefile
```

, but when I attempt to do it using a golang `http.Request.Do`, I get the above `400`. The form fields appear to be the same, so I'm thinking the culprit might have something to do with uploading a file in binary rather than using `-F file=@<filename>`.

**Comment by mitkins** [comment, marked as answer]
I faced this problem myself. Thanks to this post, it gave me the clues I was after. In the original post, there was the following data:

```
{
  url: 'http://localhost:9000/',
  fields: {
    bucket: 'bucket-name',
    'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
    'X-Amz-Credential': 'minioAccessKey/20231127/local/s3/aws4_request',
    'X-Amz-Date': '20231127T195940Z',
    key: 'uploads/d5075acc-afb7-406b-88c4-9d22eb1d8d8c/71b497a2-a74e-4fb9-8d08-9e44c65251fd.docx',
    Policy:
      'eyJleHBpcmF0aW9uIjoiMjAyMy0xMS0yN1QyMDowOTo0MFoiLCJjb25kaXRpb25zIjpbeyJidWNrZXQiOiJoYXRjaGVkLWJ5LWFpLWxvY2FsIn0seyJYLUFtei1BbGdvcml0aG0iOiJBV1M0LUhNQUMtU0hBMjU2In0seyJYLUFtei1DcmVkZW50aWFsIjoibWluaW9BY2Nlc3NLZXkxLzIwMjMxMTI3L2xvY2FsL3MzL2F3czRfcmVxdWVzdCJ9LHsiWC1BbXotRGF0ZSI6IjIwMjMxMTI3VDE5NTk0MFoifSx7ImtleSI6InVwbG9hZHMvZDUwNzVhY2MtYWZiNy00MDZiLTg4YzQtOWQyMmViMWQ4ZDhjLzcxYjQ5N2EyLWE3NGUtNGZiOS04ZDA4LTllNDRjNjUyNTFmZC5kb2N4In1dfQ==',
    'X-Amz-Signature':
      '65c14c59e7a4a09b11a2fc4ee9dc16ded298291525acd887b64c34260790ec88',
  },
}
```

I can see two problems:

* The `url` needs to be `http://localhost:9000/bucket-name`
* As far as I can tell, `X-Amz-Credential` should have an AWS region - e.g. `minioAccessKey/20231127/us-east-1/s3/aws4_request`

Another problem that was particular to my situation - I hadn't configured encryption in my Minio server. However the sample code I was using looked like this:

```elixir
  def sign_form_upload(config, bucket, opts) do
    key = Keyword.fetch!(opts, :key)
    max_file_size = Keyword.fetch!(opts, :max_file_size)
    content_type = Keyword.fetch!(opts, :content_type)
    expires_in = Keyword.fetch!(opts, :expires_in)

    expires_at = DateTime.add(DateTime.utc_now(), expires_in, :millisecond)
    amz_date = amz_date(expires_at)
    credential = credential(config, expires_at)

    encoded_policy =
      Base.encode64("""
      {
        "expiration": "#{DateTime.to_iso8601(expires_at)}",
        "conditions": [
          {"bucket":  "#{bucket}"},
          ["eq", "$key", "#{key}"],
          {"acl": "public-read"},
          ["eq", "$Content-Type", "#{content_type}"],
          ["content-length-range", 0, #{max_file_size}],
          {"x-amz-server-side-encryption": "AES256"},
          {"x-amz-credential": "#{credential}"},
          {"x-amz-algorithm": "AWS4-HMAC-SHA256"},
          {"x-amz-date": "#{amz_date}"}
        ]
      }
      """)

    fields = %{
      "key" => key,
      "acl" => "public-read",
      "content-type" => content_type,
      "x-amz-server-side-encryption" => "AES256",
      "x-amz-credential" => credential,
      "x-amz-algorithm" => "AWS4-HMAC-SHA256",
      "x-amz-date" => amz_date,
      "policy" => encoded_policy,
      "x-amz-signature" => signature(config, expires_at, encoded_policy)
    }

    {:ok, fields}
  end
```

I had to remove both references to `x-amz-server-side-encryption` - then my call was successful. I don't recommend disabling security features - but in my case this was just a test server. I just wanted to get it working!

Leaving this here in case it helps
